### ISSUE TITLE
Proposal: preload and hotstandby loading policies

### ISSUE THREAD

**Issue by db7** [lovoo/goka#156]
**Problem**: In processors with large state, scaling the number of processor instances up or down takes long because the group table partitions have to be loaded from Kafka into the local storage. While a processor instance is recovering the partitions, its assigned partitions are not processed, creating sometime a high lag.

**Proposal**: Create different loading policies to mitigate this problem. We could start with 3 policies:

- `ON_DEMAND`: When a processor instance starts, it first rebalances the partitions. The processor instance only loads the partitions assigned to it before starting to process them. This is default policy and how it currently works. 

- `PRELOAD`: When a processor instance starts, it first loads *all* partitions before doing rebalancing. After rebalance, the processor instance does not further load partitions that are not assigned to it. The local storage of processor instances will be lagging for non-assigned partitions as in `ON_DEMAND` policy.

- `HOTSTANDBY`: When a processor instance starts, it first preloads *all* partitions as in `PRELOAD` policy. After rebalance, the processor instance starts processing assigned partitions and *keeps loading* non-assigned partitions.

The `HOTSTANDBY` policy is nice because scaling down the number of processor instances or the crash of instances does not cause any lag because the remainder processor instances can start processing the dangling partitions right away after rebalance. Of course, it drastically increase the network traffic. That is a trade-off.

**Further steps**: In the future, we could introduce preload and hotstandby policies that don't load all partitions, but just a subset. That would require us to play with the rebalancing policy to guarantee that after rebalancing the instances with preloaded partitions will have a high likelihood of being assigned these partitions.

**Comment by awdng** [comment]
I was wondering if I missed something crucial in the presentation at GoDays 2019 or if the on demand scaling would actually be a problem. In scenarios with high peak loads and auto scaling, spinning up new instances that have to read and build their local state first before being ready to serve requests would be a dealbreaker in many scenarios, wouldn't it? Is it advised to use Goka in more Downstream services where processing speed is non-crucial? Should the setup be overscaled by default? I wonder what the best approach is here.

**Comment by db7** [comment]
Good questions, @awdng. Glad you are trying Goka.

> In scenarios with high peak loads and auto scaling, spinning up new instances that have to read and build their local state first before being ready to serve requests would be a dealbreaker in many scenarios

True, if your state is large, quickly scaling up will be a problem. If you are deploying on the cloud though, you can use external persistent disks for your replicas. That will mitigate the problem. If you scale down, you don't lose the complete state. Once you scale up again, the recovery will be faster (depending on your write pattern).

> Is it advised to use Goka in more Downstream services where processing speed is non-crucial?

I would say the opposite. Since you always have the state locally, your processor (and your views) won't do any hops to fetch the necessary data. That renders a very responsive system. Again, if you are running on the cloud, I'd suggest you first overscale it, then you experiment with shadow traffic and check how fast recovery can be if you keep the data in persistent disks (storage is much cheaper than compute, so that shouldn't be much of a problem).

BTW, this proposal is dealing with related issues. If you have suggestions, let me know.

**Comment by frairon** [comment]
closing this, it's been added and PR is merged.
